### PR TITLE
libcontainer: cgroups: fs: fix innerPath

### LINKED ISSUE #551: cgroup: devices: Set called with incorrect path
When vendoring runC `v0.0.8` into Docker, the following error happens:

```
docker: Error response from daemon: Cannot start container f4f299909554101cb523da3c3fa17c21204ce2a23807023a4b46dc0a1b235e09: [10] System error: write /sys/fs/cgroup/devices/devices.deny: invalid argument.
```

This is caused by `(*DevicesGroup).Set` gets given the wrong path (`/sys/fs/cgroup/devices` as opposed to the container's cgroups). I'm investigating this.

### PR DESCRIPTION
**Pull request by cyphar**
Fix m.Path legacy code to actually work. This means
we'll be able to finally vendor to Docker. It'd be nice
to merge this ASAP so we can finally merge a bunch
of stuff in Docker.

Fixes #551

Signed-off-by: Aleksa Sarai <asarai@suse.com>

/cc @crosbymichael @LK4D4

### CODE REVIEW

**Comment by cyphar** [comment]
Right, so it turns out that the patch I'm fixing also introduced a security vulnerability.

**Comment by cyphar** [comment]
The Docker suite passes once you apply this PR. Please merge ASAP. Thanks guys. :D

/ping @crosbymichael @LK4D4 @dqminh @mrunalp @vmarmol

**Comment by mlaventure** [comment]
Looks like I forgot to call the cleaning path routine in my last refactoring. Thanks for spotting it!

Could you also add a unit test for it within runc? This would avoid having to wait on vendoring into docker to find future regression if any.

**Comment by crosbymichael** [comment]
Yes, can you add some unit tests for this path logic?

**Comment by cyphar** [comment]
@crosbymichael I've added some (pretty exhaustive IMO) unit tests to make sure we don't run into this again. :P

File: libcontainer/cgroups/fs/apply_raw_test.go [hunk 1]
```diff
@@ -0,0 +1,272 @@
...
+
+// XXX: Remove me after we get rid of configs.Cgroup.Name and configs.Cgroup.Parent.
+func TestInvalidAbsoluteCgroupParent(t *testing.T) {
+	root, err := getCgroupRoot()
+	if err != nil {
+		t.Errorf("couldn't get cgroup root: %v", err)
+	}
+
+	config := &configs.Cgroup{
+		Parent: "../../../../../../../../../../some/path",
```
Comment (mlaventure) [inline libcontainer/cgroups/fs/apply_raw_test.go]: Missing the initial `/` to make it absolute

Comment (cyphar) [inline libcontainer/cgroups/fs/apply_raw_test.go]: Nit addressed.

**Comment by hqhq** [comment]
LGTM

File: libcontainer/cgroups/fs/apply_raw.go [hunk 2]
```diff
@@ -276,14 +277,19 @@ func getCgroupData(c *configs.Cgroup, pid int) (*cgroupData, error) {
...
 	}
 
-	innerPath := c.Path
+	// XXX: Do not remove this code. Path safety is important! -- cyphar
+	cgPath := libcontainerUtils.CleanPath(c.Path)
+	cgParent := libcontainerUtils.CleanPath(c.Parent)
+	cgName := libcontainerUtils.CleanPath(c.Name)
+
+	innerPath := cgPath
 	if innerPath == "" {
```
Comment (LK4D4) [inline libcontainer/cgroups/fs/apply_raw.go]: maybe just rename innerPath to cgPath to avoid temp variable

**Comment by crosbymichael** [comment]
LGTM

**Pull request merged**